core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -406,11 +407,24 @@ public static Expression convert(Expression operand, Type fromType,
     if (toPrimitive != null) {
       if (fromPrimitive != null) {
         // E.g. from "float" to "double"
-        return Expressions.convert_(
+        if (toPrimitive == Primitive.BOOLEAN) {
+          return Expressions.convert_(
+              operand, toPrimitive.getPrimitiveClass());
+        }
+        return Expressions.convertChecked(
             operand, toPrimitive.getPrimitiveClass());
       }
-      if (fromNumber || fromBox == Primitive.CHAR) {
-        // Generate "x.shortValue()".
+      if (fromType == BigDecimal.class && toPrimitive.isFixedNumeric()) {
+        // Conversion from decimal to an exact type
+        ConstantExpression zero = Expressions.constant(0);
+        // Elsewhere Calcite uses this rounding mode implicitly, so we have to be consistent.
+        // E.g., this is the rounding mode used by BigDecimal.longValue().
+        Expression rounding = Expressions.constant(RoundingMode.DOWN);
+        // rounded = operand.setScale(0, RoundingMode.DOWN);
+        Expression rounded = Expressions.call(operand, "setScale", zero, rounding);
+        // return rounded.to*ValueExact()
+        return Expressions.unboxExact(rounded, toPrimitive);
```

Comment:
conversion from decimal to some integer type needs rounding, here we use the DOWN rounding, which is implicitly used in Calcite wherever BigDecimal.longValue is used. If a different rounding mode is desired we should file an issue.